### ISSUE TITLE
[css-view-transitions-2] Support View Transitions on a DOM subtree

### ISSUE THREAD

**Issue by khushalsagar** [w3c/csswg-drafts#9890]
Currently View Transitions are Document scoped, there can only be 1 transition at a time and it runs in a layer which paints on top of the entire Document. A common feature request is to be able to run these transitions within the scope of a DOM subtree like:

```js
document.getElementById(foo).startViewTransition(updateFoo);
```

Adding Agenda+f2f to introduce rough ideas on the rendering model for this.

**Comment by bramus** [comment]
A while back I ran a poll ([Twitter](https://twitter.com/bramus/status/1726671464366211293), [Mastodon](https://front-end.social/@bramus/111444544751532812)) asking authors what is missing from View Transitions. Out of the 33 replies, 6 requested this feature, making it the number 1 request _(along with [retargetable transitions](https://github.com/w3c/csswg-drafts/issues/8687) and [auto v-t-name generation](https://github.com/w3c/csswg-drafts/issues/8320))_

**Comment by astearns** [comment]
archived presentation: https://lists.w3.org/Archives/Public/www-archive/2024Feb/0000.html

**Comment by emilio** [comment]
> Rendering and hit-testing suppression only applies to the scope

What does that mean exactly? In regular view transitions will suppress `requestAnimationFrame` and virtually all steps of the rendering loop, and display the "old" content. When rendering is suppressed on scoped view transitions, presumably that won't happen? What do you display instead?

**Comment by noamr** [comment]
> > Rendering and hit-testing suppression only applies to the scope
> 
> What does that mean exactly? In regular view transitions will suppress `requestAnimationFrame` and virtually all steps of the rendering loop, and display the "old" content. When rendering is suppressed on scoped view transitions, presumably that won't happen? What do you display instead?

@vmpstr can correct me if I'm wrong, but in broad terms the old element should render a frozen snapshot that is presented while the rest of the document can continue rendering as usual. We want to prevent a case where a lengthy callback promise on some scoped element would freeze the whole document.

**Comment by emilio** [comment]
Right, so the mechanism is completely different, correct? As in, presumably rAF / etc still need to run, and we "just" need to define something else here. Presumably the element does remain being "live" (other than not rendered...). So if it resizes or even moves then stuff gets rather weird. But I guess you solve at least the resize case by skipping...

**Comment by noamr** [comment]
> Right, so the mechanism is completely different, correct? As in, presumably rAF / etc still need to run, and we "just" need to define something else here. Presumably the element does remain being "live" (other than not rendered...). So if it resizes or even moves then stuff gets rather weird. But I guess you solve at least the resize case by skipping...

Something like this, yes. It's not too different from `iframe.contentDocument.startViewTransition` on some level.

**Comment by EisenbergEffect** [comment]
How does this work with Shadow DOM? If I'm manipulating multiple elements that are directly the children of a shadow root, what do I call `startViewTransition()` on?

**Comment by skobes-chromium** [comment]
My understanding is that scoped view transitions will conceptually operate on the flat tree and don't care about shadow DOM boundaries.  This is similar to how document view transitions can have participants inside shadow DOM ([wpt test](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/external/wpt/css/css-view-transitions/shadow-part-with-name.html)).

So you can call `startViewTransition()` on the shadow host, or one of its ancestors, and it should work as expected.

**Comment by sorvell** [comment]
To be useful in Shadow DOM you need to be able to define `view-transition-names` in a styleSheet applied to a shadowRoot. This doesn't seem to be working (see [example](https://codepen.io/Steve-Orvell/pen/EaPwNBa))?

**Comment by EisenbergEffect** [comment]
> To be useful in Shadow DOM you need to be able to define `view-transition-names` in a styleSheet applied to a shadowRoot. This doesn't seem to be working (see [example](https://codepen.io/Steve-Orvell/pen/EaPwNBa))?

This is the scenario I'm interested in seeing fully supported. I need to be able to place all relevant CSS in the Shadow DOM.

**Comment by noamr** [comment]
> How does this work with Shadow DOM? If I'm manipulating multiple elements that are directly the children of a shadow root, what do I call `startViewTransition()` on?

you can currently start the transition in an element inside the shadow scope.
But I think `shadowRoot.startViewTransition` is a feature that makes sense. Mind opening a separate issue for that?

**Comment by shonya3** [comment]
Tried to perform Scoped View Transition on el inside shadowDOM. 
And it does not work for me.
Project repo https://github.com/shonya3/scoped-view-transitions-shadow-dom-bug
Stackblitz https://stackblitz.com/~/github.com/shonya3/scoped-view-transitions-shadow-dom-bug?file=package-lock.json

**Comment by skobes-chromium** [comment]
Yes sorry about that, we are tracking a Chromium implementation bug in http://crbug.com/445481941 and it should be fixed soon.

**Comment by noamr** [comment]
Opened https://github.com/w3c/csswg-drafts/issues/12953

**Comment by martrapp** [comment]
Not sure whether I should address this here ([as requested](https://developer.chrome.com/blog/scoped-view-transitions-feedback#feedback)) or in #13123. This might need a change in Chrome Nightly, the spec, or just me being more patient while the two sync up. I would really appreciate a clarification.

In the current implementation, `view-transition-scope: auto` seems to work only on elements that actually do create a view transition group. Also, the property seems to be completely ignored in cross document view transitions. Is both of that intentional? [codepen](https://codepen.io/editor/martrapp/pen/019c2528-2748-771d-b041-90c368c7eb01)

**Comment by skobes-chromium** [comment]
Thanks for raising this. It looks like a bug in Chrome — we respect `view-transition-scope: auto` only if the element already creates a stacking context. Your codepen could work around this by setting `body { contain: layout }`. (We generally recommend for scopes to have `contain: layout`.)

**Comment by martrapp** [comment]
Thank you for your quick reply! I can confirm your suggestion that creating a stacking context did the trick. (I just added opacity: 0.99). Should I open a bug report for this?

**Comment by skobes-chromium** [comment]
I've filed [crbug.com/481934462](https://crbug.com/481934462) for this.

**Comment by martrapp** [comment]
What is the recommended way to override `view-transition-scope: all`, which is automatically applied to the `transitionRoot` of an active scoped view transition?

The [Self Participating Scopes](https://docs.google.com/document/d/1I_38aS2_W1iKT35wClsKvoJ8JYDPCmuVuhPmowSG6pU/edit?pli=1&tab=t.0) document states:

> If scopes are self-participating by default, then they should probably also be `contain: view-transition` by default, so that an outer transition does not try to use an inner transition's scope as a participant in the outer transition.

My reading of 'by default' is that it describes a default rather than a constraint, but I may be misinterpreting this. Should authors be able to override `view-transition-scope: all` in such cases, or is this deliberately not exposed? Personally, having an explicit escape hatch here would be very useful, for example, to allow the outer transition to capture elements that are added to the inner scope after the inner transition has started.

**Comment by bramus** [comment]
> Should authors be able to override `view-transition-scope: all` in such cases, or is this deliberately not exposed?

You can use `view-transition-scope: none;` in your author stylesheet to undo the effective application of `view-transition-scope: all`.

**Comment by martrapp** [comment]
> You can use `view-transition-scope: none;` in your author stylesheet to undo the effective application of `view-transition-scope: all`.

I know that was the plan, but I am probably being a bit clumsy about it.
Perhaps you could take a look at https://codepen.io/martrapp/pen/PwboOWa

**Comment by bramus** [comment]
Right, I see. I recalled using it in a demo, but that was me _adding_ `view-transition-scope: all` to elements, not undoing `view-transition-scope: none` — something that is not possible indeed.

_(https://github.com/w3c/csswg-drafts/issues/13637 is the issues in which we discussed this, so perhaps we should continue the thread there :))_
